src/librustc_lint/types.rs

```diff
@@ -427,6 +454,14 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
                                 }
                             }
                         }
+
+                        if all_phantom {
+                            if check_kind.is_toplevel() {
+                                return FfiUnsafe(UNSIZED_STRUCT_MSG);
+                            }
```

Comment:
Can we remove this `check_kind` parameter and instead pass back `FfiPhantom` (in which case a to-level caller will error out?). Or, alternatively, return `FfiPhantomWrapper` or something so that the top-level can distinguish the two cases? Threading the `check_kind` parameter doesn't seem necessary and is kind of annoying.